File: 1. 단계별로_풀어보기/20~29/21. 동적 계획법 1/5. 1912.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const N = +inputArr[0]
const numArr = inputArr[1].trim().split(' ').map(Number)

// 누적합 배열 생성
const accArr = Array.from({length: N + 1}, () => 0)

for (let i = 1; i <= N; i++) {
  accArr[i] = accArr[i - 1] + numArr[i - 1]
}

// minAcc[i] = accArr[0] ~ accArr[i - 1] 중에서 최소값
const minAcc = Array.from({length: N + 1}, () => 100000001)
for (let i = 1; i <= N; i++) {
  minAcc[i] = Math.min(minAcc[i - 1], accArr[i - 1])
}

let result = -100000001
for (let i = 1; i <= N; i++) {
  result = Math.max(result, accArr[i] - minAcc[i])
}

process.stdout.write(`${result}`)

File: 1. 단계별로_풀어보기/20~29/21. 동적 계획법 1/5. 1912.py
import sys

sys.setrecursionlimit(3 * (10 ** 5))

input = sys.stdin.readline
output = sys.stdout.write

N = int(input())
numArr = list(map(int, input().split()))

accArr = [0] * (N + 1)
minAcc = [100000001] * (N + 1)

# accArr 와 minAcc 를 구하는걸 recursion 으로 할 수 없다.
#   Python3 은 시간초과
#   Pypy3 는 메모리초과
for i in range(1, N + 1):
  accArr[i] = accArr[i - 1] + numArr[i - 1]
  minAcc[i] = min(minAcc[i - 1], accArr[i - 1])

def getResult(n):
  if n == 1:
    return accArr[n] - minAcc[n]
  
  return max(accArr[n] - minAcc[n], getResult(n - 1))

output(f"{getResult(N)}")